python_modules/dagster/dagster/core/execution_plan/objects.py

```diff
@@ -230,3 +252,42 @@ def __new__(cls, included_steps, inputs=None):
             set(check.list_param(included_steps, 'included_steps', of_type=str)),
             check.opt_dict_param(inputs, 'inputs'),
         )
+
+
+class StepOutputMap(dict):
+    def __getitem__(self, key):
+        check.inst_param(key, 'key', SolidOutputHandle)
+        return dict.__getitem__(self, key)
+
+    def __setitem__(self, key, val):
+        check.inst_param(key, 'key', SolidOutputHandle)
+        check.inst_param(val, 'val', StepOutputHandle)
+        return dict.__setitem__(self, key, val)
+
+
+# This is the state that is built up during the execution plan build process.
+# steps is just a list of the steps that have been created
+# step_output_map maps logical solid outputs (solid_name, output_name) to particular
+# step outputs. This covers the case where a solid maps to multiple steps
+# and one wants to be able to attach to the logical output of a solid during execution
+class StepBuilderState:
+    def __init__(self, pipeline_name, initial_tags=None):
+        self.steps = []
+        self.step_output_map = StepOutputMap()
+        self._tags = check.opt_dict_param(
+            initial_tags, 'initial_tags', key_type=str, value_type=str
+        )
+        self._tags['pipeline'] = pipeline_name
+
+    def get_tags(self, **additional_tags):
+        additional_tags.update(self._tags)
+        return additional_tags
+
+    @contextmanager
+    def push_tags(self, **kwargs):
```

Comment:
`tags` instead of `push_tags`?

Reply:
hmmm i think that is pretty confusing as i think of `tags` as just a getter for the tags